envoy: golang L7 enforcement infrastructure

jrajahalme opened the pull request:
This PR adds a Go interface to the `cilium.network` filter and the Go implementation of three simple test protocols exercised by running `make tests` in `/envoy`.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5174)
<!-- Reviewable:end -->

aanm commented on proxylib/npds/client.go:
```diff
@@ -0,0 +1,147 @@
+
+	"google.golang.org/genproto/googleapis/rpc/status"
+	"google.golang.org/grpc"
+
```
remove this line

jrajahalme replied:
Empty line removed.

aanm commented on proxylib/npds/client.go:
```diff
@@ -0,0 +1,147 @@
+	"google.golang.org/genproto/googleapis/rpc/status"
+	"google.golang.org/grpc"
+
+	log "github.com/sirupsen/logrus"
```
Make use of our log library with a sub-sys

jrajahalme replied:
More work is needed in formatting the logs properly. First the logs get from proxylib to Envoy logs, and from there to cilium logs. One approach would be to format them like any other Envoy logs, but we could also recognize these in cilium-agent when we receive logs from Envoy and format them just like all other Cilium logs.

Don't know yet which way to go so not working on the log formatting yet.

aanm commented on proxylib/npds/client.go:
```diff
@@ -0,0 +1,147 @@
+		var failReason error
+		for {
+			if failReason != nil {
+				log.Info(failReason)
```
`log.Info` -> `log.Error`

jrajahalme replied:
Connection errors can be transient and as we keep trying to open the connection I'm not sure these should be logged as errors.

aanm commented on proxylib/npds/client.go:
```diff
@@ -0,0 +1,147 @@
+				time.Sleep(DialBackoff)
+			}
+
+			conn, err := grpc.Dial(unixPath, grpc.WithInsecure())
```
add a comment just to explain the why `WithInsecure` option is safe

jrajahalme replied:
Added this:
```
	//
	// WithInsecure() is safe here because we are connecting to a Unix-domain socket,
	// data of whch is never on the wire and security for which can be managed with file permissions.
	//
```

aanm replied:
Perfect

aanm commented on proxylib/npds/client_test.go:
```diff
@@ -0,0 +1,77 @@
+	"github.com/cilium/cilium/proxylib/test"
+
+	"gopkg.in/check.v1"
+
```
remove space

jrajahalme replied:
Empty line removed.

aanm commented on proxylib/proxylib.go:
```diff
@@ -0,0 +1,198 @@
+	log "github.com/sirupsen/logrus"
+)
+
+var mutex lock.Mutex
```
add mutex comment
```
var(
 // mutex protects connections
)
```

jrajahalme replied:
Done!

aanm commented on proxylib/proxylib.go:
```diff
@@ -0,0 +1,198 @@
+)
+
+var mutex lock.Mutex
+var connections map[uint64]*Connection
```
add comment that what is the meaning of the `key` uint64

jrajahalme replied:
Added.

aanm commented on proxylib/proxylib.go:
```diff
@@ -0,0 +1,198 @@
+	log "github.com/sirupsen/logrus"
+)
+
+var mutex lock.Mutex
```
consider using `lock.RWMutex`?

jrajahalme replied:
Done.

aanm commented on proxylib/proxylib.go:
```diff
@@ -0,0 +1,198 @@
+// us again for the reverse direction input.
+//
+//export OnData
+func OnData(connectionId uint64, reply, endStream bool, data *[]string, filterOps *[]C.FilterOp) C.FilterResult {
```
Small nit: consider removing the pointer for the slices. Edit: one is a C array I honestly don't know what is the behavior for that.

jrajahalme replied:
While it may not be required for 'data' (which is read-only), we need a pointer for the `filterOps` slice as it's len is modified, ans if passed by value any modifications on length would remain local to the copy in stack. These slices may also never be reallocated within Go, so I wanted to be very explicit that these are allocated by C and not even the slice headers are copied.

aanm replied:
ah ok, but why don't isn't filterOps a return argument?

aanm commented on proxylib/proxylib.go:
```diff
@@ -0,0 +1,198 @@
+
+		if op == PASS || op == DROP {
+			// Skip bytes in input, or exhaust the input.
+			for bytes > 0 && unit < len(*data) {
```
you mean `len` or `cap`?

jrajahalme replied:
`len()` as in the number of strings in the slice. The `cap()` may be higher.

aanm replied:
I thought you wanted to skip all bytes, that's why I thought you wanted `cap` instead.

aanm commented on proxylib/proxylib.go:
```diff
@@ -0,0 +1,198 @@
+// us again for the reverse direction input.
+//
+//export OnData
+func OnData(connectionId uint64, reply, endStream bool, data *[]string, filterOps *[]C.FilterOp) C.FilterResult {
```
Did you consider using `data []byte` instead or `data []uint`?

jrajahalme replied:
You mean `data *[][]byte` instead of `data *[]string`?

[]string actually matches with the evbuffer scatter-gather list as string is a pointer-length pair rather than a pointer-len-cap triplet. The syntax with string is also easier and as Go strings don't get confused by zeroes it should be all-OK for binary data as well.

aanm replied:
this came from the fact that I thought you had
```
op, bytes := connection.Parser.OnData(reply, endStream, (*data)[uint:], offset)
```
instead of
```
op, bytes := connection.Parser.OnData(reply, endStream, (*data)[unit:], offset)
```

(check the `uint` vs `unit`)

aanm commented on proxylib/proxylib/connection.go:
```diff
@@ -0,0 +1,101 @@
+}
+
+// inject buffers data to be injected into the connection at the point of INJECT
+func (connection *Connection) Inject(reply bool, data []byte) int {
```
here you're not passing `data *[]byte` I wonder what make you pick one over the other in different files.

jrajahalme replied:
Here `data []byte` is memory managed by Go. This could also be `data string` for consistency, but here use of a byte slice is no problem (as it is Go allocated memory).

aanm commented on proxylib/proxylib/connection.go:
```diff
@@ -0,0 +1,101 @@
+	n := copy((*buf)[offset:cap(*buf)], data)
+	*buf = (*buf)[:offset+n] // update the buffer length
+
+	log.Debugf("proxylib: Injected %d bytes: %v (given: %s)", n, string((*buf)[offset:offset+n]), data)
```
`log.Debugf("proxylib: Injected %d bytes: %s (given: %s)", n, string((*buf)[offset:offset+n]), string(data))`

jrajahalme replied:
Some extra debugging left here, will clean up as suggested.

jrajahalme replied:
No, it was just sloppy coding ;-)

jrajahalme replied:
fixed.

aanm commented on proxylib/proxylib/connection.go:
```diff
@@ -0,0 +1,101 @@
+// connection.
+
+type Direction struct {
+	InjectBuf *[]byte
```
Does this needs to be a `*[]byte` instead of a `[]byte`?

jrajahalme replied:
The slice is allocated by C and Go may not reallocate it. We add data to these buffers from Go, and C must be aware of the added data (change in length), thus we must use the slice header provided by C, not a copy somewhere in Go memory.

aanm commented on proxylib/proxylib/connection.go:
```diff
@@ -0,0 +1,101 @@
+	Ingress    bool
+	SrcId      uint32
+	DstId      uint32
+	SrcAddr    string
```
just to be sure, it's a string because it can be a domain name as well?

jrajahalme replied:
No, these are currently always IP address/port pairs. It is just easier to deal with them as strings. These are currently only used for access logging and there addresses are strings as well. Converting specialized address types from C++ to C to Go to Protobuf is just painful when it does not provide additional value, especially so when they can be either IPv4 or IPv6.

aanm commented on proxylib/proxylib/connection.go:
```diff
@@ -0,0 +1,101 @@
+	DstId      uint32
+	SrcAddr    string
+	DstAddr    string
+	PolicyName string
```
I don't understand, where is this coming from? Can't we have multiple policies with the same name? Also, what's a policy name?

jrajahalme replied:
This could be named better, but it is the name of the policy that applies to this connection. Same policy applies to all connections to/from an endpoint. Policies are synced separately from proxy port instances so that they do not need to be repeated for all the redirects separately. In practice cilium-agent uses string formatted IP-addresses as "policy names", but they could be any identifiers that correlate a connections to a policy (npds.proto has a corresponding "name" field).

aanm commented on proxylib/proxylib/policymap.go:
```diff
@@ -0,0 +1,332 @@
+	core "github.com/cilium/cilium/pkg/envoy/envoy/api/v2/core"
+
+	"github.com/golang/protobuf/proto"
+
```
remove space

jrajahalme replied:
Done

aanm commented on proxylib/proxylib/parserfactory.go:
```diff
@@ -0,0 +1,40 @@
+	Create(connection *Connection) Parser // must be thread safe!
+}
+
+var ParserFactories map[string]ParserFactory // const after initialization
```
Consider make this private. Also you can initialize the variable here, so no need for check if it's nil on line 35. Variable initialization happens before `init()`s

jrajahalme replied:
changed as suggested; added a getter that is called from the main package.

aanm commented on proxylib/proxylib/policymap.go:
```diff
@@ -0,0 +1,332 @@
+// 'l7' interface passed by the L7 implementation to PolicyMap.Matches() as the last parameter.
+type L7RuleParser func(rule *cilium.PortNetworkPolicyRule) []L7NetworkPolicyRule
+
+var L7RuleParsers map[string]L7RuleParser // const after initialization
```
Same comments left in `ParserFactories` applies here?

jrajahalme replied:
Yes, fixed, thanks!

aanm commented on proxylib/proxylib/policymap.go:
```diff
@@ -0,0 +1,332 @@
+}
+
+type PortNetworkPolicyRule struct {
+	AllowedRemotes map[uint64]struct{}
```
What does this map maps to?

jrajahalme replied:
Nothing, it is a set of `uint64`, and is only used to figure out if the set is empty, and if not, if the remote ID `uint64` is in the set or not.

aanm commented on proxylib/proxylib/policymap.go:
```diff
@@ -0,0 +1,332 @@
+
+func (p *PortNetworkPolicyRule) Matches(remoteId uint32, l7 interface{}) bool {
+	// Remote ID must match if we have any.
+	if len(p.AllowedRemotes) > 0 {
```
no need for that, line 86 will work for nil maps as well.

jrajahalme replied:
The `len()` is tested here to find out if allowed remotes are enforced. If yes, then the remote ID must be in the set. This implements L3-dependent L7 policy enforcement.

jrajahalme replied:
So we do not want to return `false` when the map is empty.

aanm replied:
Oh ok, please add a comment with that statement here and on line 102.

aanm commented on proxylib/proxylib/policymap.go:
```diff
@@ -0,0 +1,332 @@
+			return false
+		}
+	}
+	if len(p.L7Rules) > 0 {
```
no need for that, line 92 will work for nil slices as well.

jrajahalme replied:
Same here, an empty set of L7 rules allows everything (as specified in `npds.proto`).

aanm commented on proxylib/proxylib/policymap.go:
```diff
@@ -0,0 +1,332 @@
+		if idx == 0 {
+			firstTypeName = typeName
+		} else if typeName != firstTypeName {
+			panic(fmt.Errorf("NPDS: Mismatching L7 types on the same port %v", config))
```
Does it need to be a panic?

jrajahalme replied:
Could also be more verbose error handling and returning multiple values at each level. However, we also want to recover from any panics by rejecting the policy update, so I'm using panic/recover for "exeptional" errors and normal error handling for expected errors (not handled here properly yet, but an update to this PR will properly silently ignore unknown L7 protocols, such as HTTP :-).

aanm commented on proxylib/proxylib/policymap.go:
```diff
@@ -0,0 +1,332 @@
+		} else if typeName != firstTypeName {
+			panic(fmt.Errorf("NPDS: Mismatching L7 types on the same port %v", config))
+		}
+		rules.Rules = rules.Rules[:idx+1] // Increase slice len
```
`rules.Rules = append(rules.Rules, newRule)`

jrajahalme replied:
changed.

aanm commented on proxylib/proxylib/policymap.go:
```diff
@@ -0,0 +1,332 @@
+
+		port := rule.GetPort()
+		if _, found := policy.Rules[port]; found {
+			panic(fmt.Errorf("NPDS: Duplicate port number %d in (rule: %v) (config: %v)", port, rule, config))
```
does it need to be panic?

jrajahalme replied:
See above.

aanm commented on proxylib/proxylib/policymap.go:
```diff
@@ -0,0 +1,332 @@
+		}
+
+		if rule.GetProtocol() != core.SocketAddress_TCP {
+			panic(fmt.Errorf("NPDS: Invalid transport protocol %v", rule.GetProtocol()))
```
does it need to be panic?

jrajahalme replied:
See above.

aanm commented on proxylib/proxylib/policymap.go:
```diff
@@ -0,0 +1,332 @@
+}
+
+type PortNetworkPolicies struct {
+	Rules map[uint32]PortNetworkPolicyRules
```
From context I noticed this maps a port to portNetworkpolicyrules. Why isn't `uint32` a `uint16`? Also, what if in the future we allow rules per port/protocol pair, does this map still make sense?

Also, add comment to what this map maps into.

jrajahalme replied:
This is `uint32` just because it is `uint32` in `npds.proto`. When we allow non-TCP protocols this map key definition will need to be revised, but so does the code that creates the map entries, so I don't see this as a problem for now.

jrajahalme replied:
the map maps to `PortNetworkPolicyRules` defined above.

aanm replied:
Leave a comment that `uint32` is a port number that maps to `PortNetworkPolicyRules`

aanm commented on proxylib/proxylib/policymap.go:
```diff
@@ -0,0 +1,332 @@
+	return p.Egress.Matches(port, remoteId, l7)
+}
+
+type PolicyMap map[string]*PolicyInstance
```
neewds comment to what is the key of this map

jrajahalme replied:
Added.

aanm commented on proxylib/proxylib/policymap.go:
```diff
@@ -0,0 +1,332 @@
+	}
+
+	// Store the new policy map
+	atomic.StorePointer(&policyMap, unsafe.Pointer(&newMap))
```
replace with a `setPolicyMap()`? :-)

jrajahalme replied:
Done.

aanm commented on proxylib/proxylib/policymap.go:
```diff
@@ -0,0 +1,332 @@
+}
+
+func getPolicyMap() *PolicyMap {
+	return (*PolicyMap)(atomic.LoadPointer(&policyMap))
```
`PolicyMap` doesn't need to be a pointer [1]

jrajahalme replied:
The design here is to atomically swap a map on updates. IMO the design intent is clearer when we return the pointer rather than copying the map header by value. The map is const after it has been made available, so you are right, the return value does not need to be a pointer, but so far I have preferred it to be.

jrajahalme replied:
On a second thought I'll change this to use atomic.Value instead, as in https://golang.org/pkg/sync/atomic/#Value. I added a TODO comment for now.

jrajahalme replied:
Using `atomic.Value` now and using a map directly in the apis (rather than a pointer to a map).

aanm commented on proxylib/proxylib/policymap.go:
```diff
@@ -0,0 +1,332 @@
+
+		if oldMap != nil {
+			// Locate the old version, if any
+			oldPolicy, found := (*oldMap)[policyName]
```
[1] will avoid you doing things like these `(*oldMap)[policyName]`

jrajahalme replied:
Also changed to init() the map so the old map is newer nil. Simplifies a bit.

aanm commented on proxylib/proxylib/types.go:
```diff
@@ -0,0 +1,102 @@
+type OpError uint32
+type Op struct {
+	op      uint32
+	n_bytes uint32
```
`n_bytes` of what?

jrajahalme replied:
Added this comment:
```
// The number of bytes of data the operation 'op' applies to.
```

aanm commented on proxylib/proxylib_test.go:
```diff
@@ -0,0 +1,240 @@
+	"github.com/cilium/cilium/proxylib/proxylib"
+	"github.com/cilium/cilium/proxylib/test"
+	_ "github.com/cilium/cilium/proxylib/testparsers"
+
```
remove new line

jrajahalme replied:
Done.

aanm commented on proxylib/proxylib_test.go:
```diff
@@ -0,0 +1,240 @@
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes/any"
+
```
remove new line

jrajahalme replied:
Done.

aanm commented on proxylib/test/accesslog_server.go:
```diff
@@ -0,0 +1,137 @@
+// Copyright 2017, 2018 Authors of Cilium
```
I didn't notice the other files `Copyright 2017, 2018 Authors of Cilium` -> `Copyright 2017-2018 Authors of Cilium`

jrajahalme replied:
Other files are new, this is copied and adapted for testing purposes from an existing file in the repo. Could also figure out how to refactor the code to only have one of these in the repo, but did not want to do that yet,

aanm commented on proxylib/test/accesslog_server.go:
```diff
@@ -0,0 +1,137 @@
+
+// Close removes the unix domain socket from the filesystem
+func (s *AccessLogServer) Close() {
+	s.closing = true
```
Do we need a mutex for `closing`?

and also 
```
if s.closing {
// already closing
return
}
```

jrajahalme replied:
Good catch, but I think atomic is more appropriate as it should translate to memory access only, no need to add any lock contention here.

jrajahalme replied:
Will have to use uint32 as `sync/atomic` does not support bools directly.

aanm commented on proxylib/test/tmpdir.go:
```diff
@@ -0,0 +1,31 @@
+	log "github.com/sirupsen/logrus"
+)
+
+var Tmpdir string
```
who will delete this directory when the tests are done? Consider using `c.MkDir()`

```
package main

import (
	. "gopkg.in/check.v1"
)

func (pl *ProxyLib) SetUpSuite(c *C) {
	Tmpdir := c.MkDir()
...
}
```

jrajahalme replied:
The problem with this is that the directory is deleted after the tests are done... No way to see the file contents even if the tests fail?

aanm replied:
```
go test ./... -check.work=true
```
```
-check.work - Display and do not remove the test working directory
```

It's up to you what you consider better ;-)

aanm commented on pkg/policy/l4.go:
```diff
@@ -71,13 +71,19 @@ func (l7 L7DataMap) MarshalJSON() ([]byte, error) {
 // defines all supported types of L7 parsers
 type L7ParserType string
 
+func (l7 *L7ParserType) String() string {
```
Remove the pointer from `l7 *L7ParserType`. The `String() string` interface will work for both cases

```golang
l1 := L7ParserType("foo")
l2 := L7ParserType("foo")
fmt.Printf("%s\n", l1) // foo
fmt.Printf("%s\n", &l2) // foo
```

jrajahalme replied:
Removed.

aanm commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -923,6 +935,24 @@ var (
+					"request. If omitted or empty, all requests are allowed. " +
+					"Both keys and values must be strings.",
+				Type: "object",
+				AdditionalProperties: &apiextensionsv1beta1.JSONSchemaPropsOrBool{
```
Does this schema applies to the key or value?

jrajahalme replied:
JSON object keys are always strings. AdditionalProperties allows object to have properties (fields) that are not described in the schema. The values of those fields are restricted to be strings here.

In an update that I'll push later today I have the "rule" level seen here, as in JSON and Go you can have an array of objects. I originally modeled this like the corresponding protobuf (npds.proto), where proto3 syntax does not allow repeated maps.

jrajahalme replied:
Unfortunately while OpenAPI defines `AdditionalProperties` it does not work as specified in k8s. They have wanted to provide for forwards compatibility by reserving _all_ unspecified fields to be used as specified fields in later versions of a schema, without making an older version treat newly defined fields as additional properties. This has been resolved in k8s 1.11 by allowing `AdditionalProperties` to be used when no fields are otherwise described for the object in the schema, i.e., the whole object is reserved for unspecified fields.

The use of the generic L7 policies is intended to be used as a development tool only, so restricting their use to k8s 1.11 and higher only should be no problem.

aanm commented on pkg/policy/api/rule_validation_test.go:
```diff
@@ -269,3 +269,62 @@ func (s *PolicyAPITestSuite) TestToServicesSanitize(c *C) {
+// This test ensures that PortRules using key-value pairs do not have empty keys
+func (s *PolicyAPITestSuite) TestL7Rules(c *C) {
+
+	validL7Rule := Rule{
```
Can you provide an output of how the policy looks like written in yaml?

jrajahalme replied:
I'd be happy to add another test case using yaml, but I haven't found examples from these test files on how to translate yaml to Go. Can you point me to one, e.g., for an HTTP or Kafka rule?

aanm replied:
I saw it afterwards in json, I just wanted to see human-readable format

aanm commented on pkg/policy/repository.go:
```diff
@@ -573,6 +584,8 @@ func (p *Repository) DeleteByLabels(labels labels.LabelArray) (uint64, int) {
 // JSONMarshalRules returns a slice of policy rules as string in JSON
 // representation
 func JSONMarshalRules(rules api.Rules) string {
+	log.Debugf("Marshaling rules %v", rules)
```
remove debug?

jrajahalme replied:
Done, thanks!

aanm commented on proxylib/testparsers/headerparser.policy:
```diff
@@ -0,0 +1,38 @@
+            "rules": {
+                "l7proto": "test.headerparser",
+                "l7": [
+		    {"rule": {"prefix": "bar", "contains": "beer"}},
```
I was afraid of this. Do we need to repeat  ourselves when adding a rule? It seems tow word `"rule"` repeats multiple times. Can we avoid it? Or is the intention to have more things other than "rule"?

jrajahalme replied:
Removed "rule" in my local repo, will push with these review fixes. Due to protobuf restrictions an equivalent of "rule" is needed in npds.proto, but not here or in the yaml/JSON.

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -373,169 +369,181 @@ func (e *Endpoint) regenerateBPF(owner Owner, epdir string, regenContext *Regene
-	if option.Config.DryMode {
+	// This is factored out so that we can defer the unlock, which makes the code cleaner, but also
+	// unlocks if something panics.
+	err := func() (err error) {
```
`err` -> `errFunc`?

jrajahalme replied:
This is actually a function call returning an error value set into `err`.

I'll likely separate this commit out of the PR. I needed it to debug a nil-dereference I had introduced due to a bug elsewhere.

ianvernon replied:
I agree with @aanm calling a different variable with a different type `err` is confusing, `errFunc` is much clearer.

aanm commented on proxylib/Makefile:
```diff
@@ -17,4 +19,6 @@ clean:
 
 install:
 	$(INSTALL) -m 0755 -d $(DESTDIR)$(LIBDIR)
+	pwd
```
leftover?

jrajahalme replied:
Actually still needed to debug an obscure CI issue. WIll remove once resolved.

In short: When CI builds cilium for runtime tests somehow proxylib/libcilium.so disappears before it is installed. Can not reproduce locally, even when running runtime tests with Ginkgo locally. Also not happening when building the docker images in k8s tests.

In the CI runtime test `proxylib/libcilium.so` file is there after go build, but is not there anymore just before `install`.

aanm reviewed (changes requested):
I'm the wrong person to have an opinion on this but is the `proxylib` the right directory for this? We already have `envoy` and `proxylib`. What prevents `proxylib` to live under `pkg/proxylib`?

jrajahalme reviewed (commented):
Thanks for the review!

jrajahalme committed:
envoy: Use new accesslogging format

Send HTTP access log messages using the protobuf HTTP oneof.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Add config to Cilium network filter

Currently empty.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: WriteFilter support

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Support configuration of non-http Envoy redirects

All non-HTTP Envoy redirects will be configured with a filter chain
consisting of a cilium network filter, configured to use the Cilium
Go-filter library, and Envoy's TCP proxy filter.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Generic L7 Access logging record

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
proxylib: Do not remove libcilium.so when cleaning

install can't stat `libcilium.so`, because a parallel Jenkins step
cleans up the repo before the install step. As a stop-gap fix avoid
cleaning up libcilium.so.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Key-value policy support

Add support for generic L7 policy defined with a map of key-value
pairs. It is upto the L7 protocol to define how the key-value define
the policy for the specific application protocol.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

ianvernon commented on pkg/endpoint/bpf.go:
```diff
@@ -373,169 +369,181 @@ func (e *Endpoint) regenerateBPF(owner Owner, epdir string, regenContext *Regene
-		e.Unlock()
-		err = fmt.Errorf("Unable to cache endpoint information")
-		return 0, compilationExecuted, err
+	if option.Config.DryMode {
```
Add a comment indicating why this is done here; it is not immediately obvious why we can exit in dry-mode at this time.

ianvernon commented on pkg/policy/api/l7.go:
```diff
@@ -0,0 +1,34 @@
+func (rule *PortRuleL7) Sanitize() error {
+	for k := range *rule {
+		if k == "" {
+			return fmt.Errorf("Empty key not allowed")
```
This validation is extremely limited. I'd much prefer we have some way of validating this stuff on a per-L7-protocol basis with some type of plugin architecture. This means that we now allow import of invalid rules for any L7 protocol into Cilium. The majority of the policy package assumes that this code is clean. I guess it's up to Envoy to signal down to Cilium that it can't implement the policy? How will we signal this to users?

tgraf replied:
I agree with the high-level point. The motivation for the generic map is to have a mechanism for the rapid development of new protocol parsers and to allow for experimentation. As a parser becomes mature and the configuration API stabilizes, the API can be defined and added as a well-defined type to allow for CRD validation to be performed. There is not much of a point to do much validation if it can't be done at import time as it is incredibly difficult to provide feedback at that point.

ianvernon commented on proxylib/Makefile:
```diff
@@ -6,13 +6,13 @@ DEPS := $(shell find ../pkg accesslog npds test . \( -name '*.go' ! -name '*_tes
 clean:
 	@$(ECHO_CLEAN) $(notdir $(shell pwd))
-	-$(QUIET)rm -f $(TARGET)
+#	-$(QUIET)rm -f $(TARGET)
```
Instead of this, why not some logic like the following? https://stackoverflow.com/questions/4325216/remove-all-files-except-some-from-a-directory

ianvernon reviewed (commented):
I took a brief glance at the policy and endpoint side of things. The lack of validation for the new L7 rules introduced here beyond empty keys is concerning. Please elaborate on what will happen when Cilium tries to plumb policy rules that are invalid for a given L7 protocol into Envoy.

jrajahalme committed:
proxylib: Golang L7 filter infrastructure

proxylib is a new top-level directory for Go implemtation of L7 access
control parsers that are compiled into a shared library. This shared
library is loaded at runtime by the Cilium network layer Envoy filter.

The shared library is opened once per filter chain configured to use it,
but only one copy of the library is loaded by dlopen(). Furthermore, filters
configured with the same policy protocol and access log socket paths will
share the same policy and access log connections.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
proxylib: Use ParseError() instead of panic()

Use ParseError() provided by the proxylib package instead of panic()
directly when bailing a policy update due to an parse error.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

tgraf commented on proxylib/npds/client.go:
```diff
@@ -0,0 +1,213 @@
+
+			if err != nil {
+				backOff *= 2
+				if backOff > BackOffLimit {
```
Filed https://github.com/cilium/cilium/issues/5618 as micro task to make this use the backoff package

tgraf commented on proxylib/npds/client.go:
```diff
@@ -0,0 +1,213 @@
+		// Receive next policy configuration. This will block until the
+		// server has a new version to send, which may take a long time.
+		resp, err := stream.Recv()
+		if err == io.EOF {
```
Could this return UnexpectedEOF as well?

jrajahalme replied:
Likely, will add that.

tgraf commented on proxylib/proxylib.go:
```diff
@@ -0,0 +1,227 @@
+
+		if op == PASS || op == DROP {
+			// Skip bytes in input, or exhaust the input.
+			for bytes > 0 && unit < len(*data) {
```
Non-blocking: Would be great to isolate such functionality in a function and add a unit test.

tgraf commented on proxylib/proxylib/connection.go:
```diff
@@ -0,0 +1,101 @@
+// connection.
+
+type Direction struct {
+	InjectBuf *[]byte
```
Would be great to define a type for `*[]byte` if possible

jrajahalme replied:
I defined the type `InjectBuf` and used that instead of `Direction`, as that was the only thing 
in there. Still have to use `*[]byte` in the exported API function prototype `OnNewConnection` due to cgo restrictions.

tgraf commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -923,6 +935,17 @@ var (
 	}
 
+	PortRuleL7 = apiextensionsv1beta1.JSONSchemaProps{
+		Description: "PortRuleL7 is a map of {key,value} pairs which must be present in the " +
```
Non-blocking: I think the description is a bit too specific. We may use this for things other than policy. I would reword this to a generic statement that says that the map is passed to the parser referenced in l7proto and it is up to the parser to define what to do with the map data.

jrajahalme replied:
Changed, but kept the note that "If omitted or empty, all requests are allowed", as that is part of the matching logic that is independent of the L7. This allows cilium-agent to create wildcard policies when needed without understanding the specifics of a L7 protocol and is also the way HTTP policies work.

Currently the L7 implementation has to call the matching logic, though, so a parser could be written to violate this rule.

tgraf commented on pkg/policy/api/rule_validation.go:
```diff
@@ -259,12 +258,32 @@ func (pr *L7Rules) sanitize() error {
+	}
+	if pr.L7Proto != "" {
+		if pr.L7 == nil {
+			return fmt.Errorf("'l7' must be specified when 'l7proto' is specified")
```
Non-blocking. Do we really need to enforce this? We may want to provide parsers which do not require configuration at all.

jrajahalme replied:
Removed the nil check as range on nil slice works as intended.

tgraf commented on pkg/policy/api/utils.go:
```diff
@@ -64,6 +64,30 @@ func (k *PortRuleKafka) Equal(o PortRuleKafka) bool {
 }
 
+// Exists returns true if the L7 rule already exists in the list of rules
+func (h *PortRuleL7) Exists(rules L7Rules) bool {
```
Non-blocking: unit test

jrajahalme replied:
Added.

tgraf commented on pkg/policy/api/utils.go:
```diff
@@ -64,6 +64,30 @@ func (k *PortRuleKafka) Equal(o PortRuleKafka) bool {
+}
+
+// Equal returns true if both L7 rules are equal
+func (h *PortRuleL7) Equal(o PortRuleL7) bool {
```
Non-blocking: unit test

jrajahalme replied:
Added (in a new PR)

tgraf commented on pkg/policy/l4.go:
```diff
@@ -121,6 +123,8 @@ func (l7 L7DataMap) GetRelevantRules(identity *identity.Identity) api.L7Rules {
 			if selector.Matches(identity.Labels.LabelArray()) {
 				rules.HTTP = append(rules.HTTP, endpointRules.HTTP...)
 				rules.Kafka = append(rules.Kafka, endpointRules.Kafka...)
+				rules.L7Proto = endpointRules.L7Proto // XXX: Need to check types are the same?
```
The conflict management is done in `mergeL4IngressPort()`

jrajahalme replied:
Right, removing the comment.

tgraf commented on pkg/policy/rule.go:
```diff
@@ -85,7 +85,7 @@ func mergeL4IngressPort(ctx *SearchContext, endpoints []api.EndpointSelector, en
 			switch {
 			case len(newL7Rules.HTTP) > 0:
-				if len(ep.Kafka) > 0 {
+				if len(ep.Kafka) > 0 || len(ep.L7) > 0 {
```
This does not seem to detect a conflict if l7proto is set but the l7 map is empty. I would check for `ep.L7Proto != ""`.

jrajahalme replied:
Right, in the same vain I'm changing the (new) case to test for the `L7Proto` rather than an non-empty `L7`

jrajahalme replied:
Noticed that I had not updated mergeL4EgressPort() at all. It has a lot of duplicate code, will factor it out into a helper to remove the duplication.

tgraf commented on proxylib/testparsers/headerparser.go:
```diff
@@ -67,11 +67,10 @@ func (rule *HeaderRule) Matches(data interface{}) bool {
 	l7Rules := rule.GetL7Rules()
 	if l7Rules == nil {
-		panic(fmt.Errorf("Can't get L7 rules."))
+		ParseError("Can't get L7 rules.", rule)
```
Thank you
